KAFKA-15649: Handle directory failure timeout

viktorsomogyi opened the pull request:
A broker that is unable to communicate with the controller will shut down after the configurable log.dir.failure.timeout.ms.

The implementation adds a new event to the Kafka EventQueue. This event is deferred by the configured timeout and will execute the shutdown if the heartbeat communication containing the failed log dir is still pending with the controller.

An extra test is added to LogDirFailureTest to cover the case when a failure couldn't be communicated with the broker. The test case shuts down the controller and causes a log dir failure to assert the expected broker shutdown.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

viktorsomogyi committed:
KAFKA-15649: Handle directory failure timeout

A broker that is unable to communicate with the controller will shut down
after the configurable log.dir.failure.timeout.ms.

The implementation adds a new event to the Kafka EventQueue. This event
is deferred by the configured timeout and will execute the shutdown
if the heartbeat communication containing the failed log dir is still
pending with the controller.

viktorsomogyi commented:
Rebased it due to conflicts.

soarez commented on server/src/main/java/org/apache/kafka/server/config/Defaults.java:
```diff
@@ -94,6 +94,7 @@ public class Defaults {
     public static final int LOG_FLUSH_START_OFFSET_CHECKPOINT_INTERVAL_MS = 60000;
     public static final int NUM_RECOVERY_THREADS_PER_DATA_DIR = 1;
     public static final boolean AUTO_CREATE_TOPICS_ENABLE = true;
+    public static final long LOG_DIR_FAILURE_TIMEOUT_MS = 30000L;
```
This default seems reasonable to me.

soarez commented on core/src/main/scala/kafka/server/BrokerLifecycleManager.scala:
```diff
@@ -507,6 +522,7 @@ class BrokerLifecycleManager(
         if (errorCode == Errors.NONE) {
           val responseData = message.data()
           failedAttempts = 0
+          offlineDirs = offlineDirs.map(kv => kv._1 -> true)
```
I think this is incorrect. If a new failed directory is added to `offlineDirs` in-between a hearbeat request-resopnse, then we'll clear it here before knowing if it will propagated to the controller.

One idea is to hand down the offline dirs set in the request in `sendBrokerHeartBeat()` to `BrokerHeartbeatResponseEvent` through `BrokerHeartbeatResponseHandler` as a new constructor argument.

soarez commented on core/src/main/scala/kafka/server/BrokerLifecycleManager.scala:
```diff
@@ -327,16 +333,25 @@ class BrokerLifecycleManager(
+  private class OfflineDirBrokerFailureEvent(offlineDir: Uuid) extends EventQueue.Event {
+    override def run(): Unit = {
+      if (!offlineDirs.getOrElse(offlineDir, false)) {
+        error(s"Shutting down because couldn't communicate offline log dirs with controllers")
```
We should include the directory in the error. It might also be helpful to resolve the directory ID to its path. Perhaps something like `dirIdToPath` in `AssignmentsManager` should be made available here as well.

viktorsomogyi replied:
I'll print the UUID here only and I'll modify other log statements to contain the UUID so one can pair these log statements when analyzing the logs. Printing the dir path here would be a little bit bigger stretch as we currently don't propagate it down to this level. Let me know if you think it'd be better to print the path here.

soarez commented on core/src/main/scala/kafka/server/BrokerServer.scala:
```diff
@@ -211,7 +211,8 @@ class BrokerServer(
         isZkBroker = false,
-        logDirs = logManager.directoryIdsSet)
+        logDirs = logManager.directoryIdsSet,
+        () => kafkaScheduler.schedule("shutdown", () => shutdown(), 0, -1))
```
There's a `scheduleOnce` alternative which sets `periodMs` to `-1`.

soarez commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -528,6 +529,10 @@ object KafkaConfig {
     "this specified threshold. This configuration is ignored if log.message.timestamp.type=LogAppendTime."
 
+  val LogDirFailureTimeoutMsDoc = "If the broker is unable to successfully communicate to the controller that some log " +
+    "directory has failed for longer than this time, and there's at least one partition with leadership on that directory, " +
```
> and there's at least one partition with leadership

We aren't checking for this condition. We can either a) implement it; or b) keep it simple and drop this out of the configuration description.

viktorsomogyi replied:
I'll do another round with this, there might be a way in `BrokerServer` to extract this information using the combination of `MetadataCache`, `ReplicaManager` and `LogManager`. I'll update you tomorrow about my findings.

mimaison commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -870,6 +875,7 @@ object KafkaConfig {
       .define(CreateTopicPolicyClassNameProp, CLASS, null, LOW, CreateTopicPolicyClassNameDoc)
       .define(AlterConfigPolicyClassNameProp, CLASS, null, LOW, AlterConfigPolicyClassNameDoc)
       .define(LogMessageDownConversionEnableProp, BOOLEAN, LogConfig.DEFAULT_MESSAGE_DOWNCONVERSION_ENABLE, LOW, LogMessageDownConversionEnableDoc)
+      .define(LogDirFailureTimeoutMsProp, LONG, Defaults.LOG_DIR_FAILURE_TIMEOUT_MS, atLeast(0), MEDIUM, LogDirFailureTimeoutMsDoc)
```
In the KIP the accepted value range is defined as >= 1. I wonder if values below 1s actually make much sense. 
Also the importance was defined as low.

viktorsomogyi committed:
Address code review comments

viktorsomogyi commented:
Rebased on latest trunk as there were some conflicts. Addressed some of the comments but there are 2 things I need to investigate:
* `LogDirFailureTest` fails in `@AfterAll` likely because an incorrect shutdown, perhaps there's a timing issue
* Check if we can detect if there are any leaders before shutdown
I'll update on both shortly, hopefully tomorrow.

viktorsomogyi committed:
Avoid using the scheduler to get out of infinte loop during shutdown

viktorsomogyi commented:
@soarez at the end I chose the shortcut regarding detecting leaders before shutdown. The reason is complex as the solution that would be required for this is complex too.
So on one part the sequence of events is problematic. First we update the `LogManager` and then try to propagate the event to the controller. At this point the metadata is stale so I can't use that for reliable information to detect whether partitions have leadership or not. A workaround would be to subtract the LogManager's data from metadata cache (ie. if there is only a single isr replica and that is the current, then we can accept it as offline in reality). I don't really feel that it is a robust solution, it could be prone to race conditions on the network depending on how requests come from the controller as long as it's alive. I think it's more robust to just fail if we can't contact the controller.
The second reason is a bit technical and can be worked around, although requires lots of effort. When trying to extract which replica->logdir information from `LogManager`, my only available information regarding logdirs given by the event is the `Uuid`. Unfortunately `LogManager` doesn't store the `Uuid` of an offline dir (and besides I don't think `Uuid` and logdir names used consistently across the whole module). This problem can be solved by propagating both logdir and `Uuid` in the events or store offline dirs' `Uuid ` in `LogManager`. I think the latter is problematic because we can't know the point until we should store information about offline dirs as they might never come back. The first can be done, although could be a sizeable refactor and generally I felt that just choosing the simpler route now could be more robust.
Let me know if you think we should try it.

soarez commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -2466,7 +2467,6 @@ class ReplicaManager(val config: KafkaConfig,
            s"for partitions ${partitionsWithOfflineFutureReplica.mkString(",")} because they are in the failed log directory $dir.")
     }
     // retrieve the UUID here because logManager.handleLogDirFailure handler removes it
```
We should move or remove this comment now that the `uuid` declaration has been moved up.

soarez reviewed (commented):
@viktorsomogyi I think you're right. I'd also add a third reason to keep this simple, which is that this is a mitigation for a problem which at this point is only theoretical.

If proved necessary in the future we can go down the more complex route. Regarding the first problem you describe, the broker knows – regardless of stale metadata – which partitions it is effectively leading. And in case of a race where a lagging metadata update promotes it to a leader for some partition, that can be dealt with by extending `maybeUpdateTopicAssignment` to handle assignments into directories that are offline, correcting them to `DirectoryId.LOST`.

I think this looks good I just have a tiny comment.

viktorsomogyi committed:
Move UUID comment up

viktorsomogyi commented:
@soarez thanks for the info. I addressed your comment. Do you have anything more to add or are we good to go?

soarez reviewed (approved):
Thanks for pinging me @viktorsomogyi.

All the failed tests in CI pass locally.

LGTM
